rfcs/001-mid-level-file-api.md

```diff
@@ -0,0 +1,187 @@
+# Mid-Level File Api
+
+## Summary
+
+This is a mid-level API wrapper around the various File related JavaScript apis. The possibility for a higher level API is left open.
+
+The mid-level file API aims to implement File I/O on top of the raw JS apis found here: https://w3c.github.io/FileAPI/ which includes listing files, creating files, and reading files.
+
+## The API
+
+First we have `FileList` which is simply a list of `File`s. This is an opaque struct that offers a way to iterate over `Files` and get individual files by index:
+
+```rust
+#[derive(Debug, Clone)]
+struct FileList { ... }
+
+impl FileList {
+  fn get(index: usize) -> Option<File> { ... }
+
+  fn len(&self) -> usize { ... }
+
+  fn iter(&self) -> FileListIter { ... }
+
+  fn as_raw(&self) -> &web_sys::FileList { ... }
+}
+
+impl AsRef<web_sys::FileList> for FileList { ... }
+impl From<web_sys::FileList> for FileList { ... }
+impl Index<usize> for FileList {
+    type Output = File;
+    fn index(&self, index: usize) -> &Self::Output { ... }
+}
+```
+
+There is no `FileList::new` since creating a raw `web_sys::FileList` is not possible without going through a `web_sys::HtmlInputElement`.
+
+Next we have a `blob` module that includes the trait `BlobLike`
+```rust
+trait BlobLike {
+    fn size(&self) -> u64 { ... }
```

Comment:
This definitely shouldn't be `u64`, since `u64` cannot be sent to/from JS (it causes an exception).

So it either needs to be `f64` or `u32`. I'm not sure which one.

Reply:
This is the Rust-facing API, not what is transferred at the ABI level. It would be perfectly valid to transfer the size as an `f64` at the ABI level and then expose it to Rust as a `u64`. This gives us more bits than a `u32` and also lets us ignore silliness about fractional bytes, all while working within the ABI's contstraints.

Reply:
That's an interesting solution, I hadn't thought of that.

I think we need to be a little careful around the limits with regard to `f64`, and it needs to documented, but it seems doable.